src/utils/data_loader.py

```diff
@@ -0,0 +1,196 @@
+    data_by_item_id__train: SerialUnidirectionalMapper
+    data_by_user_id__test: SerialUnidirectionalMapper
+    data_by_item_id__test: SerialUnidirectionalMapper
+    # Mapping data
+    id_to_user_bmap: SerialBidirectionalMapper
+    id_to_item_bmap: SerialBidirectionalMapper
+
+
+class AbstractDataLoader(ABC):
+
+    @abstractmethod
+    def load(self):
+        pass
+
+
+class DataLoader(AbstractDataLoader):
+    """
+    We want to load the data as an optimized form of sparse matrices
+    """
+
+    LIMIT_TO_LOAD_IN_MEMORY = 1_000_000_000_000
+
+    DataLoadingError = type("InvalidStateError", (Exception,), {})
+
+    def __init__(
+        self,
+        *,
+        file_path: str,
+        user_header: str,
+        item_header: str,
+        rating_header: str,
+        data_headers: Iterable = None,
+        data_constructor: Callable = None,
+        verbose=True,
+        limit: int = None,
+    ):
+        self._user_header = user_header
+        self._item_header = item_header
+        self._rating_header = rating_header
+        self._file_path = file_path
+        self._data_headers = data_headers
+        self._data_constructor = data_constructor
+        self._verbose = verbose
+        self._limit = limit or self.LIMIT_TO_LOAD_IN_MEMORY
+
+    def _construct_data(self, data, items):  # noqa
+        """
+        Add structure to data
+        """
+        return (
+            self._data_constructor(data, items)
+            if self._data_constructor is not None
+            else data
+        )
+
+    def load(self, approximate_train_ratio: float = 1) -> LoadedDataWrapper:
+
+        assert (
+            0 <= approximate_train_ratio <= 1
+        ), f"Expected `train_ratio` to be in [0, 1] but got {approximate_train_ratio}"
+
+        logger.warning(
+            "The current implementation does not split the data into train and test sets exactly with the provided ration. "
+            "We use the provided ratio as a probability for a Bernoulli distribution to know either the data point should "
+            "be used as a training data or a test data."
+        )
+
+        indexed_count = 0
+
+        # The two next bmap variables are needed to keep track of the bijective
+        # mapping between each user or item and the id we've associated to them.
+        id_to_item_bmap = SerialBidirectionalMapper()  # bijective mapping
+        id_to_user_bmap = SerialBidirectionalMapper()
+
+        # This two next variables are used to group data by item_id and user_id
+        data_by_user_id__train = SerialUnidirectionalMapper()  # subjective mapping
+        data_by_item_id__train = SerialUnidirectionalMapper()
+
+        data_by_user_id__test = SerialUnidirectionalMapper()  # subjective mapping
+        data_by_item_id__test = SerialUnidirectionalMapper()
+
+        belongs_to_test_split: Optional[bool | object]
+        rand_trace = []
+        try:
+            with open(self._file_path, mode="r", newline="") as csvfile:
```

Comment:
- [ ] benchmark this code against "one using numpy" and another one using memory

Reply:
Done, the actual implementation remains the fasters for now (without jax or numba things)